servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleConcatWithPublisher.java

```diff
@@ -90,28 +96,94 @@ public final void onSubscribe(final Subscription subscription) {
 
         @Override
         public final void onNext(@Nullable final T t) {
+            // propagateCancel - if cancel does subscribe to the Publisher there will be no demand propagated upstream
+            // so we don't have to worry about concurrency or use-after-terminate here.
             target.onNext(t);
         }
 
         @Override
         public final void onError(final Throwable t) {
-            target.onError(t);
+            if (propagateCancel) {
+                onErrorPropagateCancel(t);
+            } else {
+                target.onError(t);
+            }
+        }
+
+        private void onErrorPropagateCancel(Throwable t) {
+            for (;;) {
+                final Object oldValue = mayBeResult;
+                if (oldValue == CANCELLED) {
+                    // CANCELLED means downstream doesn't care about being notified and internal state tracking also
+                    // uses this state if terminal has been propagated, so avoid duplicate terminal. We may
+                    // subscribe to both sources in parallel if cancellation occurs, so allowing terminal to
+                    // propagate would mean ordering and concurrency needs to be accounted for between Single and
+                    // Publisher, because cancel allows for no more future delivery we avoid future invocation of
+                    // the target subscriber.
```

Comment:
This is similar to https://github.com/apple/servicetalk/pull/2372#discussion_r984911440, but this operator has a lot more states. Will it be possible to make behavior consistent?

Reply:
we can propagate error as long as `propagateCancel == false` to avoid concurrency and ordering issues.